### PR TITLE
feat(core): Disambiguate Interlok/Adaptris Message metadata interface methods

### PR DESCRIPTION
**Pull request by quotidian-ennui**
## Motivation

Fixes https://github.com/adaptris/interlok/issues/946 so that we now get correct logging. But the scope changed to refactor some of the underlying InterlokMessage / AdaptrisMessage metadata interface methods.

## Modification

- Deprecate `InterlokMessage.setMessageHeaders(Map)` with javadoc change to describe the reality that it was never a set even though contractually it should have been.
- Deprecate `AdaptrisMessage.setMetadata(Set)` because it's poorly named
- Add `InterlokMessage.addMessageHeaders(Map)` and `AdaptrisMessage.addMetadata(Collection)` as default methods
- Add `InterlokMessage.replaceAllMessageHeaders(Map)` and `AdaptrisMessage.replaceAllMetadata(Collection)` as default methods.
- Add `InterlokMessage.clearMessageHeaders()`
- Mark all existing deprecated methods for removal in 5.0 in AdaptrisMessage, they've been deprecated since 3.0.6 !!
- Use hashset.clear() when clearing metadata.
- Change the behaviour of `AdaptrisMessageImp.clone()` so that it doesn't rely on non-obvious behaviour from clearMetadata() - (it relied on a new hashset being created).

## PR Checklist

- [x] been self-reviewed.
- [x] Added javadocs for most classes and all non-trivial methods
- [x] Added unit tests or modified existing tests to cover new code paths

## Result

- Better logging.
- No effect on the UI.

## Testing

Add a service that copies metadata, and check the logging.

### CODE REVIEW

**Comment by sebastien-belin-adp** [comment]
Should we do the same for setMessageHeaders (from InterlokMessage interface) as it only adds metadata as well?

**Comment by sebastien-belin-adp** [comment]
Should we do the same for setMessageHeaders (from InterlokMessage interface) as it only adds metadata as well?

**Comment by quotidian-ennui** [comment]
> Should we do the same for setMessageHeaders (from InterlokMessage interface) as it only adds metadata as well?

I think setMessageHeaders was basically a c/p from the AM interface. I think you're probably right; in fact I'd propose then we do this:

- deprecate setMessageHeaders, setMetadata
- add addMessageHeaders(Map), addMetadata(Collection)  (which can be defaulted).
- add replaceAllMessageHeaders(Map), replaceAllMetadata(Set<>) (calls clear() then add())

Is probably the right move just to clear up ambiguity.

**Comment by sebastien-belin-adp** [comment]
> 

Sounds like a plan

**Comment by quotidian-ennui** [comment]
Interestingly, DefaultSerializableMessage does obey its contract...

File: interlok-core/src/main/java/com/adaptris/core/AdaptrisMessageImp.java [hunk 1]
```diff
@@ -197,42 +196,29 @@ public synchronized void addMetadata(MetadataElement e) {
...
-  public synchronized void setMetadata(Set<MetadataElement> set) {
-    if (set != null) {
-      for (MetadataElement e : set) {
-        addMetadata(e);
-      }
-    }
-  }
-
-  @Override
+  // The behavoiur of setMessageHeaders() was always additive, so we
```
[the remark this answers is not in the document]
Comment (quotidian-ennui) [reply interlok-core/src/main/java/com/adaptris/core/AdaptrisMessageImp.java:211]: Lame correcting my spelling in a // comment ;)

**Comment by quotidian-ennui** [comment]
@sebastien-belin-adp -> because this is an internally breaking change, if something doesn't extend AdaptrisMessageImp they're going to be "broken" (i.e. they implement InterlokMessage or AdaptrisMessage directly).

IIRC interlok-swiftmq might have had a custom message type which I think has gone. I did a search in the adaptris org for "implements AdaptrisMessage" and "implements InterlokMessage" and found nothing obvious so we might be OK.

**Comment by sebastien-belin-adp** [comment]
> @sebastien-belin-adp -> because this is an internally breaking change, if something doesn't extend AdaptrisMessageImp they're going to be "broken" (i.e. they implement InterlokMessage or AdaptrisMessage directly).
> 
> IIRC interlok-swiftmq might have had a custom message type which I think has gone. I did a search in the adaptris org for "implements AdaptrisMessage" and "implements InterlokMessage" and found nothing obvious so we might be OK.
> 
> https://github.com/search?q=org%3Aadaptris+%22implements+AdaptrisMessage%22&type=code

I'll check the other interlok optionals but we should be fine indeed.

**Pull request merged**